l10n_ch_scan_bvr/wizard/scan_bvr.py

```diff
@@ -345,21 +345,52 @@ def validate_bvr_string(self):
         #
         data = {}
         data['bvr_struct'] = self._get_bvr_structurated(self.bvr_string)
+        partner_bank_model = self.env['res.partner.bank']
+        partner_bank = partner_bank_model
         # We will now search the account linked with this BVR
         if data['bvr_struct']['domain'] == 'name':
             domain = [('acc_number', '=', data['bvr_struct']['beneficiaire'])]
+            partners_bank = partner_bank_model.search(domain)
+            partner_bank = (
+                partners_bank and len(partners_bank) == 1 and
+                partners_bank[0])
         else:
-            domain = [
-                ('bvr_adherent_num', '=', data['bvr_struct']['bvrnumber'])]
-        partners_bank = self.env['res.partner.bank'].search(domain)
+            # A postal account that refers to a bank is uniquely identified
+            # by (beneficiaire + bvrnumber), at least by beneficiaire
+            domain = []
+            cond_1 = (
+                'acc_number', '=', data['bvr_struct']['beneficiaire'])
+            domain.append(cond_1)
+            partners_bank = partner_bank_model.search(domain)
+            if len(partners_bank) == 1:
+                if (
+                        not partners_bank[0].bvr_adherent_num or
+                        partners_bank[0].bvr_adherent_num ==
+                        data['bvr_struct']['bvrnumber']):
+                    partner_bank = partners_bank[0]
+            if len(partners_bank) > 1:
+                # We need to filter further by bvr_adherent_num
+                cond_2 = (
+                    'bvr_adherent_num', '=',
+                    data['bvr_struct']['bvrnumber'])
+                domain.append(cond_2)
+                partners_bank = partner_bank_model.search(domain)
+                partner_bank = (
+                    partners_bank and len(partners_bank) == 1 and
+                    partners_bank[0])
+                if len(partners_bank) > 1:
+                    raise UserError(
+                        _('There are more than one bank corresponding '
+                          'to the current string.\n'
+                          'Please check the banks configuration.'))
```

Comment:
To filter you can use `filtered` to search in the records you already fetched from database.
Plus inverse check to do only one check. Plus you can do the assignation of partner_bank only after all checks.
What would be even better is to do this in a method that returns partner_bank.

    # remove line 370 partner_bank = partners_bank[0]
    ... 
        # We need to filter further by bvr_adherent_num
        partners_bank.filtered(lambda rec: rec.bvr_adherent_num == data['bvr_struct']['bvrnumber'])
        if len(partners_bank) > 1:
            raise UserError(
               _('There are more than one bank corresponding '
                 'to the current string.\n'
                 'Please check the banks configuration.'))
    partner_bank = partners_bank